src/Sylius/Bundle/CoreBundle/Doctrine/ORM/ProductRepository.php

```diff
@@ -29,13 +29,17 @@ class ProductRepository extends BaseProductRepository implements ProductReposito
      */
     public function createByTaxonPaginator(TaxonInterface $taxon, array $criteria = [])
     {
+        $root = $taxon->isRoot() ? $taxon : $taxon->getRoot();
+
         $queryBuilder = $this->createQueryBuilder('o');
         $queryBuilder
             ->innerJoin('o.taxons', 'taxon')
+            ->andWhere($queryBuilder->expr()->eq('o.root', ':root'))
```

Comment:
As far as I can tell they are not enough. You can have left and right value which match the condition from above but the node actually have a different root. I came up with this fix because in sylius admin we found some products being listed under a taxon even they were never assigned to it. Of course it's not easy to replicate this issue but it happens.